assets/model_monitoring/components/src/action_analyzer/action_analyzer_relevance_action_detector/relevance_action_detector.py

```diff
@@ -0,0 +1,606 @@
+    """Get the query topic."""
+    idx = 0
+    if len(good_topics_dict) == 0 and row[DOCUMENT_RELEVANCE_SCORE_COLUMN] >= GOOD_METRICS_VALUE:
+        return DEFAULT_TOPIC_NAME
+    if len(bad_topics_dict) == 0 and row[DOCUMENT_RELEVANCE_SCORE_COLUMN] < METRICS_VIOLATION_THRESHOLD:
+        return DEFAULT_TOPIC_NAME
+
+    for topic, q_list in bad_topics_dict.items():
+        if row[MODIFIED_PROMPT_COLUMN] in q_list:
+            # do not show the topic name when disabling the conf
+            topic = topic if llm_summary_enabled == "true" else f"topic_{idx}"
+            return topic
+        idx += 1
+    for topic, q_list in good_topics_dict.items():
+        if row[MODIFIED_PROMPT_COLUMN] in q_list:
+            # do not show the topic name when disabling the conf
+            topic = topic if llm_summary_enabled == "true" else f"topic_{idx}"
+            return topic
+        idx += 1
+    return None
+
+
+def get_index_id(index_content):
+    """Parse the index id from index yaml."""
+    index_payload = yaml.safe_load(index_content)
+    # if the asset id does not exist, use the index name
+    if "self" in index_payload:
+        index_id = index_payload["self"].get("asset_id", None)
+    elif "index" in index_payload:
+        index_id = index_payload["index"].get("index", None)
+    else:
+        index_id = None
+    return index_id
+
+
+def parse_debugging_info(root_span):
+    """Parse the span tree to get debugging info."""
+    try:
+        tree = SpanTree.create_tree_from_json_string(root_span)
+        spans_array = []
+        prompt_flow_input = tree.root_span.input
+        for span in tree:
+            if span.span_type == RETRIEVAL_SPAN_TYPE:
+                parent_id = span.parent_id
+                if not parent_id:
+                    print("No look up span found, skip action analyzer.")
+                    return None
+                index_span = tree.get_span_tree_node_by_span_id(parent_id)
+                index_input = json.loads(json.loads(index_span.attributes)["inputs"])
+                index_content = index_input['mlindex_content']
+                retrieval_query_type = index_input["query_type"]
+                retrieval_top_k = index_input["top_k"]
+                index_id = get_index_id(index_content)
+                retrieval_info = json.loads(span.attributes)
+                query = retrieval_info["retrieval.query"]
+                retrieval_documents = json.loads(retrieval_info["retrieval.documents"])
+                text = []
+                score = []
+                for document in retrieval_documents:
+                    text.append(document["document.content"])
+                    score.append(float(document["document.score"]))
+                spans_array.append((span.span_id, index_content, index_id, query, TEXT_SPLITTER.join(text), max(score), retrieval_query_type, retrieval_top_k, prompt_flow_input))  # noqa
```

Comment:
you may want to define a class instead of a tuple here. it is hard to maintain on line 368+